artifact/scripts/executor.cpp

```diff
@@ -0,0 +1,252 @@
+	HandlerFunction handler) {
+	// Collect the error and carry on
+	if (err.code
+		== common::processes::MakeError(common::processes::NonZeroExitStatusError, "").code) {
+		this->error_script_error_ = this->error_script_error_.FollowedBy(executor::error::MakeError(
+			executor::error::NonZeroExitStatusError,
+			"Got non zero exit code from script: " + *current_script));
+	} else {
+		this->error_script_error_ = this->error_script_error_.FollowedBy(err);
+	}
+
+	// Schedule the next script execution
+	auto local_err = Execute(std::next(current_script), end, handler);
+	if (local_err != merror::NoError) {
+		return handler(local_err);
+	}
+}
+
+void ScriptRunner::HandleScriptError(Error err, HandlerFunction handler) {
+	if (err.code
+		== common::processes::MakeError(common::processes::NonZeroExitStatusError, "").code) {
+		if (this->script_.get()->GetExitStatus() == 21) {
+			return handler(executor::error::MakeError(
+				executor::error::RetryExitCodeError, "Received exit code 21"));
+		}
+		return handler(executor::error::MakeError(
+			executor::error::NonZeroExitStatusError,
+			"Received error code: " + to_string(this->script_.get()->GetExitStatus())));
+	}
+	return handler(err);
+}
+
+Error ScriptRunner::Execute(
+	vector<string>::iterator current_script,
+	vector<string>::iterator end,
+	HandlerFunction handler) {
+	// No more scripts to execute
+	if (current_script == end) {
+		handler(this->error_script_error_); // Success
+		return merror::NoError;
+	}
+
+	log::Info("Running Artifact script: " + *current_script);
+
+	this->script_.reset(new mender::common::processes::Process({*current_script}));
+	auto err {this->script_->Start(stdout_callback_, stderr_callback_)};
+	if (err != merror::NoError) {
+		return err;
+	}
+
+	return this->script_.get()->AsyncWait(
+		this->loop_,
+		[this, current_script, end, handler](Error err) {
+			if (err != merror::NoError) {
+				if (this->action_ == Action::Error) {
+					return HandleErrorScriptError(err, current_script, end, handler);
+				}
+				return HandleScriptError(err, handler);
+			}
+			// Schedule the next script execution
+			auto local_err = Execute(std::next(current_script), end, handler);
+			if (local_err != merror::NoError) {
+				return handler(local_err);
+			}
+		},
+		chrono::seconds {this->timeout_});
+}
+
+Error ScriptRunner::AsyncRunScripts(HandlerFunction handler) {
+	if (this->is_artifact_script_) {
+		// Verify the version in the version file (OK if no version file present)
+		auto version_file_error {CorrectVersionFile(
+			path::Join(common::conf::paths::DefaultArtScriptsPath(), "version"))};
+		if (version_file_error != merror::NoError) {
+			return version_file_error;
+		}
+	}
+
+	// Collect
+	if (!this->scripts_) {
+		return executor::error::MakeError(
+			executor::error::Code::CollectionError,
+			"Failed to get the scripts, error: " + this->scripts_.error().String());
+	}
+
+	// Sort
+	sort(this->scripts_.value().begin(), this->scripts_.value().end());
```

Comment:
`auto &scripts = this->scripts_.value()` could make things here and below a bit nicer.